### PR TITLE
[SPARK-43205][SQL][FOLLOWUP] add ExpressionWithUnresolvedIdentifier to simplify code

### PR DESCRIPTION
**Pull request by cloud-fan**
### What changes were proposed in this pull request?

This is a followup of https://github.com/apache/spark/pull/41007 . It adds `ExpressionWithUnresolvedIdentifier` to simplify the code, following the same idea of `PlanWithUnresolvedIdentifier`

### Why are the changes needed?

code cleanup

### Does this PR introduce _any_ user-facing change?

no

### How was this patch tested?

existing tests

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @gengliangwang @srielau

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveIdentifierClause.scala [hunk 1]
```diff
@@ -18,39 +18,51 @@
...
           case null => expr.failAnalysis(
             errorClass = "NOT_A_CONSTANT_STRING.NULL",
             messageParameters = Map(
-              "name" -> clauseName,
+              "name" -> "IDENTIFIER",
               "expr" -> expr.sql))
-          case other => other.toString // OK
+          case other =>
+            // Parse the identifier string to name parts.
+            UnresolvedAttribute(other.toString).nameParts
```
Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveIdentifierClause.scala:64]: This one is a bit confusing. Shall we just use 
```
CatalystSqlParser.parseMultipartIdentifier(name)
```

File: sql/core/src/test/resources/sql-tests/analyzer-results/ansi/double-quoted-identifiers-disabled.sql.out [hunk 2]
```diff
@@ -218,8 +218,8 @@ org.apache.spark.sql.AnalysisException
     "objectType" : "",
     "objectName" : "",
     "startIndex" : 15,
-    "stopIndex" : 37,
-    "fragment" : "not_exist AS X(`hello`)"
+    "stopIndex" : 23,
```
Comment (gengliangwang) [inline sql/core/src/test/resources/sql-tests/analyzer-results/ansi/double-quoted-identifiers-disabled.sql.out:221]: Awesome, the context is accurate now.

**Review by gengliangwang — APPROVED** [review]
This simplifies the code, thanks for the work!

**Comment by cloud-fan** [comment]
thanks for the review, merging to master!

**Pull request closed without merging**